### PR TITLE
Make the first party extensions optional, add [extensions] extra

### PR DESCRIPTION
**Pull request by hroncok**
This is my attempt to make all the sphinxcontrib requirements optional.

### Feature or Bugfix

- Feature

### Purpose

We package sphinx in Fedora Linux and we don't really have a strong reason to pull in packages like applehelp there.
I decided to see if we could get rid of them.

### Detail

<del>Note that this is a very early draft. I am yet to test this in Fedora and see if this approach is feasible.</del> It works for us.

If it is, this obviously requires some documentation changes and possible deprecations. I am however unable to drive that forward right now. I would be interested to hear from the Sphinx maintainers whether this approach makes sense or not at all.

### Relates
- https://github.com/sphinx-doc/sphinx/issues/11567

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -76,6 +70,24 @@ dependencies = [
...
+    "sphinxcontrib-htmlhelp>=2.0.0",
+]
+serializinghtml = [
+    "sphinxcontrib-serializinghtml>=1.1.9",
+]
+qthelp = [
+    "sphinxcontrib-qthelp",
+]
 docs = [
     "sphinxcontrib-websupport",
```
Comment (hroncok) [inline pyproject.toml:100]: TODO maybe add the dependencies here?

**Comment by hroncok** [comment]
I have identified 5 Fedora packages that are impacted by this change. Most of them explicitly build the docs in some of the removed formats and are fixed by declaring a dependency on the particular sphinxcontrib package. One of them was a bit more interesting, as the tests only (accidentally) succeeded with `sphinxcontrib_jsmath-1.0.1-py3.12-nspkg.pth` insatlled: https://github.com/fastavro/fastavro/pull/730

Long story short, we would like to have this change in Fedora. Please let me know if we should carry this downstream-only or if you are eventually interested as well.

**Comment by hroncok** [comment]
The flake8 CI fail in a file I have not changed.

**Comment by AA-Turner** [comment]
This would be too painful a change in terms of deprecations at the moment, so I've closed this PR.

A
